src/utilities/cache-eviction-policy.ts

```diff
@@ -0,0 +1,139 @@
+ * of an item release it, that item is considered evictable.
+ */
+export class CacheEvictionPolicy<T = string> {
+  private[$retainerCount] = new Map<T, number>();
+  private[$recentlyUsed]: Array<T> = [];
+  private[$evictionThreshold]: number;
+  private[$cache]: MutableCache<T>;
+
+  constructor(cache: MutableCache<T>, evictionThreshold: number = 5) {
+    this[$cache] = cache;
+    this[$evictionThreshold] = evictionThreshold;
+  }
+
+  /**
+   * The eviction threshold is the maximum number of items to hold
+   * in cache indefinitely. Items within the threshold (in recently
+   * used order) will continue to be cached even if they have zero
+   * retainers.
+   */
+  set evictionThreshold(value: number) {
+    this[$evictionThreshold] = value;
+    this[$evict]();
+  }
+
+  get evictionThreshold(): number {
+    return this[$evictionThreshold];
+  }
+
+  /**
+   * A reference to the cache that operates under this policy
+   */
+  get cache(): MutableCache<T> {
+    return this[$cache];
+  }
+
+  /**
+   * Given an item key, returns the number of retainers of that item
+   */
+  retainerCount(key: T): number {
+    return this[$retainerCount].get(key) || 0;
+  }
+
+  /**
+   * Resets the internal tracking of cache item retainers. Use only in cases
+   * where it is certain that all retained cache items have been accounted for!
+   */
+  reset() {
+    this[$retainerCount].clear();
+    this[$recentlyUsed] = [];
+  }
+
+  /**
+   * Mark a given cache item as retained, where the item is represented
+   * by its key. An item can have any number of retainers.
+   */
+  retain(key: T) {
+    if (!this[$retainerCount].has(key)) {
+      this[$retainerCount].set(key, 0);
+    }
+    this[$retainerCount].set(key, this[$retainerCount].get(key)! + 1);
+
+    const recentlyUsedIndex = this[$recentlyUsed].indexOf(key);
+
+    if (recentlyUsedIndex !== -1) {
+      this[$recentlyUsed].splice(recentlyUsedIndex, 1);
+    }
+
+    this[$recentlyUsed].unshift(key);
+    // Evict, in case retaining a new item pushed an evictable item beyond the
+    // eviction threshold
+    this[$evict]();
+  }
+
+  /**
+   * Mark a given cache item as released by one of its retainers, where the item
+   * is represented by its key. When all retainers of an item have released it,
+   * the item is considered evictable.
+   */
+  release(key: T) {
+    if (this[$retainerCount].has(key)) {
+      this[$retainerCount].set(
+          key, Math.max(this[$retainerCount].get(key)! - 1, 0));
+    }
+
+    this[$evict]();
+  }
+
+  [$evict]() {
+    if (this[$recentlyUsed].length < this[$evictionThreshold]) {
+      return;
+    }
+
+    for (let i = this[$recentlyUsed].length - 1; i >= this[$evictionThreshold];
+         --i) {
+      const key = this[$recentlyUsed][i];
+      const retainerCount = this[$retainerCount].get(key);
+
+      if (retainerCount === 0) {
+        this[$cache].delete(key);
+        this[$recentlyUsed].splice(i, 1);
```

Comment:
I don't quite follow this splice. Why are we putting a new entry into $recentlyUsed when we just deleted that entry from the cache?

Reply:
[The signature for JavaScript splice](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice) is:

```
var arrDeletedItems = array.splice(start[, deleteCount[, item1[, item2[, ...]]]])
```

Added items start at the third argument. So, we are only removing things here, not adding anything.

Reply:
Ah, that's what I missed when I read that definition; I thought these splices were adding elements, but now I see they're removing them. I thought elements weren't being removed at all, hence my other comment. Thanks!